drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,2019 @@
+	 * IN ACK ISR of SET_ADDRESS control transfer.
+	 */
+	data->addr = addr;
+
+	numaker_usbd_unlock(dev);
+
+	return 0;
+}
+
+void usb_dc_set_status_callback(const usb_dc_status_callback cb)
+{
+	const struct device *dev = numaker_usbd_device_get();
+	struct numaker_usbd_data *data = dev->data;
+
+	numaker_usbd_lock(dev);
+
+	data->status_cb = cb;
+
+	numaker_usbd_unlock(dev);
+}
+
+int usb_dc_ep_check_cap(const struct usb_dc_ep_cfg_data *const ep_cfg)
+{
+	const struct device *dev = numaker_usbd_device_get();
+	const struct numaker_usbd_config *config = dev->config;
+	int rc = 0;
+	struct numaker_usbd_ep *ep_cur;
+
+	numaker_usbd_lock(dev);
+
+	/* For safe, require EP number for control transfer to be 0 */
+	if ((ep_cfg->ep_type == USB_DC_EP_CONTROL) && USB_EP_GET_IDX(ep_cfg->ep_addr) != 0) {
+		LOG_ERR("EP number for control transfer must be 0");
+		rc = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Some soc series don't allow ISO IN/OUT to be assigned the same EP number.
+	 * This is addressed by limiting all OUT/IN EP addresses in top/bottom halves,
+	 * except CTRL OUT/IN.
+	 */
+	if (config->disallow_iso_inout_same && ep_cfg->ep_type != USB_DC_EP_CONTROL) {
+		/* Limit all OUT EP addresses in top-half, except CTRL OUT */
+		if (USB_EP_DIR_IS_OUT(ep_cfg->ep_addr) && USB_EP_GET_IDX(ep_cfg->ep_addr) >= 8) {
+			LOG_DBG("Support only ISO OUT EP address 0x01~0x07: 0x%02x",
+				ep_cfg->ep_addr);
+			rc = -ENOTSUP;
+			goto cleanup;
+		}
+
+		/* Limit all IN EP addresses in bottom-half , except CTRL IN */
+		if (USB_EP_DIR_IS_IN(ep_cfg->ep_addr) && USB_EP_GET_IDX(ep_cfg->ep_addr) < 8) {
+			LOG_DBG("Support only ISO IN EP address 0x88~0x8F: 0x%02x",
+				ep_cfg->ep_addr);
+			rc = -ENOTSUP;
+			goto cleanup;
+		}
+	}
+
+	/* To respect this capability check, pre-bind EP context to EP address,
+	 * and pre-determined its type
+	 */
+	ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep_cfg->ep_addr);
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep=0x%02x", ep_cfg->ep_addr);
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+cleanup:
+
+	numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+int usb_dc_ep_set_callback(const uint8_t ep, const usb_dc_ep_callback cb)
+{
+	const struct device *dev = numaker_usbd_device_get();
+	int rc = 0;
+	struct numaker_usbd_ep *ep_cur;
+
+	numaker_usbd_lock(dev);
+
+	/* Bind EP context to EP address */
+	ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep);
+
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep=0x%02x", ep);
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+	ep_cur->cb = cb;
+
+cleanup:
+
+	numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data *const ep_cfg)
+{
+	const struct device *dev = numaker_usbd_device_get();
+	int rc = 0;
+	uint32_t dmabuf_base;
+	uint32_t dmabuf_size;
+	struct numaker_usbd_ep *ep_cur;
+
+	LOG_DBG("Configure: ep=0x%02x, ep_mps=%d, ep_type=%d", ep_cfg->ep_addr, ep_cfg->ep_mps,
+		ep_cfg->ep_type);
```

Comment:
It does not look nice because you trust stupid tools to format the code and do not check it.
```c
	LOG_DBG("Configure: ep=0x%02x, ep_mps=%d, ep_type=%d",
		ep_cfg->ep_addr, ep_cfg->ep_mps, ep_cfg->ep_type);
```

Reply:
yes, i am relying on `clang-format` to format this code. however, as pointed out, it doesn't do well on some line breaking. currently, for manual refine, i use`/* clang-format on */` and `/* clang-format off */` directives to avoid roll-back by `clang-format`.  it seems zephyr allows this coding style.

Reply:
I do not accept any patches containing /* clang-format on */ or /* clang-format off */. This is your tooling problem and you can setup/use editor that has better clang-format integration that allows you to format only desired regions of the code.

Reply:
removed `/* clang-format on */` and `/* clang-format off */`. reduced the line to in 100-chars limit so that `clang-format` won't reformat it.